ThirdPartyAdapters/vungle/vungle/src/main/java/com/vungle/mediation/VungleBannerAdapter.java

```diff
@@ -0,0 +1,265 @@
+                new VungleInitializer.VungleInitializationListener() {
+                    @Override
+                    public void onInitializeSuccess() {
+                        loadBanner();
+                    }
+
+                    @Override
+                    public void onInitializeError(String errorMessage) {
+                        Log.d(TAG, "SDK init failed:" + VungleBannerAdapter.this);
+                        mIsLoading.set(false);
+                        mVungleManager.removeActiveBannerAd(mPlacementId);
+                        if (mVungleListener != null) {
+                            mVungleListener.onAdFailedToLoad(AdRequest.ERROR_CODE_INTERNAL_ERROR);
+                        }
+                    }
+                });
+    }
+
+    void destroy(@Nullable View adView) {
+        Log.d(TAG, "Vungle banner adapter try to destroy.");
+        if (adView == mAdLayout) {
+            Log.d(TAG, "Vungle banner adapter destroy:" + this);
+            mIsLoading.set(false);
+            mVisibility = false;
+            mPendingRequestBanner = false;
+            mVungleManager.removeActiveBannerAd(mPlacementId);
+            if (mVungleBannerAd != null) {
+                mVungleBannerAd.destroyAd();
+                mVungleBannerAd = null;
+            }
+            if (mVungleNativeAd != null) {
+                mVungleNativeAd.finishDisplayingAd();
+                mVungleNativeAd = null;
+            }
+        }
+    }
+
+    void cleanUp() {
+        Log.d(TAG, "Vungle banner adapter try to cleanUp.");
+        if (mVungleBannerAd != null) {
+            Log.d(TAG, "Vungle banner adapter cleanUp: destroyAd");
+            mVungleBannerAd.destroyAd();
+            mVungleBannerAd = null;
+        }
+        if (mVungleNativeAd != null) {
+            Log.d(TAG, "Vungle banner adapter cleanUp: finishDisplayingAd");
+            mVungleNativeAd.finishDisplayingAd();
+            mVungleNativeAd = null;
+        }
+    }
+
+    void preCache() {
+        if (AdConfig.AdSize.isBannerAdSize(mAdConfig.getAdSize())) {
+            Banners.loadBanner(mPlacementId, mAdConfig.getAdSize(), null);
+        } else {
+            Vungle.loadAd(mPlacementId, null);
+        }
+    }
+
+    void updateVisibility(boolean visible) {
+        this.mVisibility = visible;
+        if (mVungleBannerAd != null) {
+            mVungleBannerAd.setAdVisibility(visible);
+        } else if (mVungleNativeAd != null) {
+            mVungleNativeAd.setAdVisibility(visible);
+        }
+    }
+
+    private LoadAdCallback mAdLoadCallback = new LoadAdCallback() {
+        @Override
+        public void onAdLoad(String id) {
+            createBanner();
+        }
+
+        @Override
+        public void onError(String id, VungleException exception) {
+            Log.d(TAG, "Ad load failed:" + VungleBannerAdapter.this);
+            mIsLoading.set(false);
+            mVungleManager.removeActiveBannerAd(mPlacementId);
```

Comment:
@HuFangshuai have you tested noUUID scenario?
Seems the second banner gets expected error but finishes the first one, and second will stuck in play operatioons

Similar is reproducible for 1 banner, on fail (by timeout waiting from admob) we still will create & hold VungleBanner in map and never finish it

Reply:
For none UUID scenario:
1.banner1 is loading, banner2 will throw away.(map size=1)
2.banner1 loaded success, banner2's request will be identified as banner1's refresh. If banner2 load failure, will finish banner1 ad instance.(map size=0)
Do you mean case 2?

Reply:
fixed. by put to map when the last request loaded(VungleBannerAdapter#createBanner).